library/src/main/java/org/apache/apex/malhar/lib/dedup/AbstractDeduper.java

```diff
@@ -0,0 +1,462 @@
+
+  /**
+   * Map to hold the result of a tuple processing (unique, duplicate, expired or error) until previous
+   * tuples get processed. This is used only when {@link #orderedOutput} is true.
+   */
+  private transient Map<T, Decision> decisions;
+  private transient long sleepMillis;
+  private transient Map<T, Future<Slice>> waitingEvents = Maps.newLinkedHashMap();
+  private transient Map<Slice, Long> asyncEvents = Maps.newLinkedHashMap();
+
+  // Metrics
+  @AutoMetric
+  private transient long uniqueEvents;
+  @AutoMetric
+  private transient long duplicateEvents;
+  @AutoMetric
+  private transient long expiredEvents;
+
+  @Override
+  public void setup(OperatorContext context)
+  {
+    sleepMillis = context.getValue(Context.OperatorContext.SPIN_MILLIS);
+    FileAccessFSImpl fAccessImpl = new TFileImpl.DTFileImpl();
+    fAccessImpl.setBasePath(context.getValue(DAG.APPLICATION_PATH) + "/bucket_data");
+    managedState.setFileAccess(fAccessImpl);
+    managedState.setup(context);
+
+    if (orderedOutput) {
+      decisions = Maps.newLinkedHashMap();
+    }
+  }
+
+  @Override
+  public void beginWindow(long l)
+  {
+    // Reset Metrics
+    uniqueEvents = 0;
+    duplicateEvents = 0;
+    expiredEvents = 0;
+
+    managedState.beginWindow(l);
+  }
+
+  protected abstract Slice getKey(T event);
+
+  protected abstract long getTime(T event);
+
+  /**
+   * Processes an incoming tuple
+   *
+   * @param tuple
+   */
+  protected void processTuple(T tuple)
+  {
+
+    long time = getTime(tuple);
+    Future<Slice> valFuture = managedState.getAsync(time, getKey(tuple));
+
+    if (valFuture.isDone()) {
+      try {
+        processEvent(tuple, valFuture.get());
+      } catch (InterruptedException | ExecutionException e) {
+        throw new RuntimeException("process", e);
+      }
+    } else {
+      processWaitingEvent(tuple, valFuture);
+    }
+  }
+
+  /**
+   * Processes a looked-up event
+   *
+   * @param tuple
+   * @param value
+   */
+  protected void processEvent(T tuple, Slice value)
+  {
+    if (value == BucketedState.EXPIRED) {
+      processInvalid(tuple);
+      return;
+    }
+    processValid(tuple, value);
+  }
+
+  /**
+   * Processes a tuple which is waiting for the lookup to return.
+   *
+   * @param tuple The tuple which needs to wait
+   * @param future The future object which will ultimately return the lookup result
+   */
+  protected void processWaitingEvent(T tuple, Future<Slice> future)
+  {
+    waitingEvents.put(tuple, future);
+    if (orderedOutput) {
+      recordDecision(tuple, Decision.UNKNOWN);
+    }
+  }
+
+  /**
+   * Processes a valid (non-expired) tuple. This tuple may be a unique or a duplicate.
+   *
+   * @param tuple
+   *          The tuple to be processed
+   * @param value
+   *          Looked up key of the tuple
+   */
+  protected void processValid(T tuple, Slice value)
+  {
+    if (!orderedOutput || waitingEvents.isEmpty()) {
```

Comment:
We are checking whether we can directly proceed to process a tuple and emit it. If we don't need to preserve order, then we can proceeed. Also even if we need to preserve the order, and there is no other event that is waiting; then we can proceed.